Building DLL with --noMain produces an unexpected DllMain on devel branch

### Description

If we try to compile the following code:

```nim
import std/winlean

proc NimMain() {.cdecl, importc.}

proc DllMain*(hModule: Handle; reasonForCall: DWORD; reserved: pointer): WINBOOL {.stdcall, exportc, dynlib.} =
  if reasonForCall == 1:
    NimMain()
  
  return 1
```

Using the following command:

```cmd
nim c --app:lib --noMain mymodule.nim 
```

We will get a compilation error reporting a type conflict between two **DllMain** definitions as if we had not provided **--noMain**.
If we try to compile the same code with the same command on a stable branch, there is no error and the DLL will compile.


### Nim Version

Nim Compiler Version 1.9.5 [Windows: amd64]
Compiled at 2023-07-22
Copyright (c) 2006-2023 by Andreas Rumpf   

git hash: https://github.com/nim-lang/Nim/commit/62869a5c68e4dd91e00ee77b039f0175482ef4fa
active boot switches: -d:release

### Current Output

```text
C:\Users\User\nimcache\mymodule_d\@mmymodule.nim.c:123:13: error: conflicting types for 'DllMain'; have 'BOOL(struct HINSTANCE__ *, DWORD,  void *)' {aka 'int(struct HINSTANCE__ *, long unsigned int,  void *)'}
  123 | BOOL WINAPI DllMain(HINSTANCE hinstDLL, DWORD fwdreason,
      |             ^~~~~~~
In file included from C:\Users\User\nimcache\mymodule_d\@mmymodule.nim.c:7:
C:\Users\User\nimcache\mymodule_d\@mmymodule.nim.c:69:30: note: previous definition of 'DllMai 
' with type 'NI32(NI,  NI32,  void *)' {aka 'int(long long int,  int,  void *)'}
   69 | N_LIB_EXPORT N_STDCALL(NI32, DllMain)(NI hModule_p0, NI32 reasonForCall_p1, void* reserved_p2) {      
      |                              ^~~~~~~
C:\Users\User\.choosenim\toolchains\nim-#devel\lib/nimbase.h:167:54: note: in definition of macro 'N_STDCALL'
  167 | #  define N_STDCALL(rettype, name) rettype __stdcall name
      |                                                      ^~~~
Error: execution of an external compiler program 'gcc.exe -c  -w -fmax-errors=3 -mno-ms-bitfields -DWIN32_LEAN_AND_MEAN   -IC:\Users\User\.choosenim\toolchains\nim-#devel\lib -IC:\nim-dllmain -o C:\Users\User\nimcache\mymodule_d\@mmymodule.nim.c.o C:\Users\User\nimcache\mymodule_d\@mmymodule.nim.c' failed with exit code: 1
```


### Expected Output

```text
Build DLL
```


### Possible Solution

_No response_

### Additional Information

This started happening on the devel branch build dated **2023-05-27** and supposedly caused by the commit https://github.com/nim-lang/Nim/commit/609bf3d7c8bf880dea70f6e8211976f3ec15

--- a/compiler/cgen.nim
+++ b/compiler/cgen.nim
@@ -1636,7 +1636,7 @@ proc genMainProc(m: BModule) =
     appcg(m, m.s[cfsProcs], nimMain,
         [m.g.mainModInit, initStackBottomCall, m.labels, preMainCode, m.config.nimMainPrefix, isVolatile])
 
-  if optNoMain notin m.config.globalOptions or optGenDynLib in m.config.globalOptions:
+  if optNoMain notin m.config.globalOptions:
     if m.config.cppCustomNamespace.len > 0:
       closeNamespaceNim(m.s[cfsProcs])
       m.s[cfsProcs].add "using namespace " & m.config.cppCustomNamespace & ";\L"

--- a/compiler/commands.nim
+++ b/compiler/commands.nim
@@ -802,7 +802,6 @@ proc processSwitch*(switch, arg: string, pass: TCmdLinePass, info: TLineInfo;
       defineSymbol(conf.symbols, "consoleapp")
     of "lib":
       incl(conf.globalOptions, optGenDynLib)
-      incl(conf.globalOptions, optNoMain)
       excl(conf.globalOptions, optGenGuiApp)
       defineSymbol(conf.symbols, "library")
       defineSymbol(conf.symbols, "dll")

